### ISSUE TITLE
Query:Iter() does not timeout in some cases

### ISSUE THREAD

**Issue by zhixinwen** [gocql/gocql#896]
We set the timeout in each query to be 500 ms, but sometimes it can take more than that.

It becomes a trouble when we have one slow Cassandra node due to network issue recently, and we see around 5% of the queries return using more than 1.5 seconds.

**Comment by zhixinwen** [comment]
This issue may be related to #839

**Comment by Zariel** [comment]
This is odd because the timeout for the query is done independent of IO
whilst waiting for a response. Query timeout does not include parsing the
result, what total time are you seeing? Can you instrument the code and see
where the time is being spent?

**Comment by zhixinwen** [comment]
I would add more log when I have time.
The issue is mostly obvious when one of C* node is behaving slow.
For a 500ms request, I can see response time from 800ms to 10s.

**Comment by Zariel** [comment]
the timeout is per request to cassandra, so if you read 10 pages the total timeout will be 10 * timeout, you can use a context to prevent multiple pages from spanning over a timeout.

**Comment by zhixinwen** [comment]
I set both timeout and the context timeout to be 500 ms.

**Comment by zhixinwen** [comment]
I spend some time digging into the code and here is the possible root cause.
```
       // conn.go
	err := req.writeFrame(framer, stream)
	if err != nil {
		// closeWithError will block waiting for this stream to either receive a response
		// or for us to timeout, close the timeout chan here. Im not entirely sure
		// but we should not get a response after an error on the write side.
		close(call.timeout)
		// I think this is the correct thing to do, im not entirely sure. It is not
		// ideal as readers might still get some data, but they probably wont.
		// Here we need to be careful as the stream is not available and if all
		// writes just timeout or fail then the pool might use this connection to
		// send a frame on, with all the streams used up and not returned.
		c.closeWithError(err)
		return nil, err
	}
```
`req.writeFrame` would call:
```
func (c *Conn) Write(p []byte) (int, error) {
	if c.timeout > 0 {
		c.conn.SetWriteDeadline(time.Now().Add(c.timeout))
	}

	return c.conn.Write(p)
}
```
Because of the semantic of `conn.SetWriteDeadline`, it would extend the ddl every time a request reuses a connection.
It is fine during normal time, but when C* is slow, it would indefinitely extend the ddl. In practice, I have seen the result returns in >30s

**Comment by zhixinwen** [comment]
Instead of writing to connection directly in `conn.Write()`, I am planning to write to a buffered channel and write to connection asynchronously.

I would do some testing and send out the patch later.

Please let me know, if you have any concern over the way.

**Comment by Zariel** [comment]
That makes sense

**Comment by zhixinwen** [comment]
@Zariel I planned to send the request to the buffered channel (size is configurable). If the channel is full, directly return an error and rely on read side to handle the timeout. 
If we do not use a buffered channel, I am afraid `conn.Write() ` would still be possible to get stuck when sending to channel. Of cause, we can discard the write if we would get stuck when sending to the channel, but it maybe too aggressive if the network is just temporarily slow.

**Comment by Zariel** [comment]
I was thinking it would would look something like, correct me if i'm wrong.

```go
select {
case <-c.writeFrame(f):
case <-timeout:
// handle
case <- ctx.Done():
// handle
}

select {
case <-call.resp:
case <-timeout:
}
```

This can open the way to have separate read + write timeouts and overall deadlines.

**Comment by zhixinwen** [comment]
In this case, `c.writeFrame(f)` needs to run in a separate goroutine, otherwise it can still get stuck? I would send out a pull request with my original idea of patch, and we can see how that works.

**Comment by Zariel** [comment]
Yeah I was imagining we would have a goroutine that just writes frames and another that just reads them.

**Comment by zhixinwen** [comment]
I mean `c.writeFrame(f)` seems to need to spawn a new goroutine per request instead of per connection. Maybe I miss something in between.

**Comment by Zariel** [comment]
my bad, I meant something like

```
select {
case c.writeFrame <- f:
case <-timeout:
// handle
case <- ctx.Done():
// handle
}
```

**Comment by zhixinwen** [comment]
Your proposed way can work, but when I try to implement it, I find it hard to come up with a clean interface. I think https://github.com/gocql/gocql/pull/1061 would also work, and we can depend on `timeoutch` and `ctxDone` to do the timeout. Let me know if you have any comments.

**Comment by martin-sucha** [comment]
It seems that a fix for this was merged in #1062 .
